File: Blind 75/13 - 1D Dynamic Programming/647-palindromic-substrings.py
'''
Given a string s, return the longest palindromic substring in s.
'''
from time import time


class Solution:
    def countSubstrings(self, s: str) -> int:
        res = 0
        # Odds
        for i in range(len(s)):
            j = i
            while i >= 0 and j < len(s) and s[i] == s[j]:
                res += 1
                i -= 1
                j += 1

        # Evens
        for i in range(len(s)):
            j = i + 1
            while i >= 0 and j < len(s) and s[i] == s[j]:
                res += 1
                i -= 1
                j += 1

        return res

    def countPali(self, s, l, r):
        res = 0
        while l >= 0 and r < len(s) and s[l] == s[r]:
            res += 1
            l -= 1
            r += 1
        return res

    def reference(self, s: str) -> int:
        res = 0

        for i in range(len(s)):
            res += self.countPali(s, i, i)
            res += self.countPali(s, i, i + 1)
        return res

    def quantify(self, test_cases, runs=100000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.countSubstrings(case))
                else:
                    self.countSubstrings(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        'abc',
        'aaa'
    ]
    test.quantify(test_cases)

File: Blind 75/13 - 1D Dynamic Programming/JavaScript/647-palindromic-substrings.js
const { performance } = require('perf_hooks');

class Solution {
  method(s) {
    let count = 0;

    for (let i = 0; i < s.length; i++) {
      let [l, r] = [i, i];
      while (
        l >= 0 &&
        l < s.length &&
        r >= 0 &&
        r < s.length &&
        s[l] === s[r]
      ) {
        l -= 1;
        r += 1;
        count += 1;
      }
    }

    for (let i = 0; i < s.length; i++) {
      let [l, r] = [i, i + 1];
      while (
        l >= 0 &&
        l < s.length &&
        r >= 0 &&
        r < s.length &&
        s[l] === s[r]
      ) {
        l -= 1;
        r += 1;
        count += 1;
      }
    }

    return count;
  }

  reference(s, count = 0) {
    for (let i = 0; i < s.length; i++) {
      /* Time O(N) */
      const [odd, even] = [i, i + 1];
      /* odd-length: single character center */
      count += this.isPalindromeFromCenter(s, i, odd); /* Time O(N) */
      /* even-length: consecutive characters center */
      count += this.isPalindromeFromCenter(s, i, even); /* Time O(N) */
    }

    return count;
  }

  isPalindromeFromCenter = (s, left, right, count = 0) => {
    const isInBounds = () => 0 <= left && right < s.length;
    while (isInBounds()) {
      /* Time O(N) */
      const isEqual = s[left] === s[right];
      if (!isEqual) break;

      count++;

      left--;
      right++;
    }

    return count;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = ['abc', 'aaa'];
test.quantify(testCases);
